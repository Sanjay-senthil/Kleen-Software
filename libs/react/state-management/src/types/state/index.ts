export * from './endUser';
export * from './globalAppErrors';
export * from './investigation';
