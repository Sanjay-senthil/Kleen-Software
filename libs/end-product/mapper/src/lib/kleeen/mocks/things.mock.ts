import { ActionType, StatisticalDataType, ThingMap } from '@kleeen/types';

export const thingsMock: ThingMap = {
  completedService: {
    actions: [
      {
        areYouSure: true,
        component: 'CompletedServiceReportIssueModal',
        description: undefined,
        displayName: `Report Issue`,
        id: `4d882b7c-3773-443f-9327-16d168155e3c`,
        isCustomModal: true,
        name: `reportIssue`,
        type: ActionType.Custom,
      },
    ],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `dateScheduled`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceRecordType`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `downtimeHours`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `estimatedCost`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `datePerformed`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceCost`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceDescription`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `uuid`,
    deepDataType: undefined,
    description: ``,
    displayName: `CompletedService`,
    editable: false,
    format: { examples: [] },
    formatType: `uuid`,
    id: 59317,
    multiple: true,
    name: `completedService`,
    prototypeId: 59314,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.FreeForm,
    type: `uuid`,
  },
  currentMileage: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `integer`,
    deepDataType: undefined,
    description: ``,
    displayName: `CurrentMileage`,
    editable: false,
    format: { min: 0, max: 189000, examples: [] },
    formatType: `integer`,
    id: 59309,
    multiple: true,
    name: `currentMileage`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.NumericDiscrete,
    type: `integer`,
  },
  datePerformed: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `serviceRecord`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `date`,
    deepDataType: undefined,
    description: ``,
    displayName: `DatePerformed`,
    editable: false,
    format: { examples: [] },
    formatType: `date`,
    id: 59303,
    multiple: true,
    name: `datePerformed`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.NumericTemporal,
    type: `date`,
  },
  dateScheduled: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `serviceRecord`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `date`,
    deepDataType: undefined,
    description: ``,
    displayName: `DateScheduled`,
    editable: false,
    format: { examples: [] },
    formatType: `date`,
    id: 59308,
    multiple: true,
    name: `dateScheduled`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.NumericTemporal,
    type: `date`,
  },
  downtimeHours: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `serviceRecord`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `integer`,
    deepDataType: undefined,
    description: ``,
    displayName: `DowntimeHours`,
    editable: false,
    format: { min: 1, max: 24, examples: [] },
    formatType: `integer`,
    id: 59305,
    multiple: true,
    name: `downtimeHours`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.NumericDiscrete,
    type: `integer`,
  },
  estimatedCost: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `serviceRecord`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `integer`,
    deepDataType: undefined,
    description: ``,
    displayName: `EstimatedCost`,
    editable: false,
    format: { min: 1, max: 100, prefix: '$', suffix: 'k', examples: [] },
    formatType: `USD`,
    id: 59304,
    multiple: true,
    name: `estimatedCost`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.NumericDiscrete,
    type: `integer`,
  },
  scheduledService: {
    actions: [
      {
        areYouSure: true,
        component: 'ScheduledServiceReportIssueModal',
        description: undefined,
        displayName: `Report Issue`,
        id: `4d882b7c-3773-443f-9327-16d168155e3c`,
        isCustomModal: true,
        name: `reportIssue`,
        type: ActionType.Custom,
      },
    ],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `dateScheduled`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceRecordType`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `downtimeHours`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `estimatedCost`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `datePerformed`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceCost`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceDescription`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `uuid`,
    deepDataType: undefined,
    description: ``,
    displayName: `ScheduledService`,
    editable: false,
    format: { examples: [] },
    formatType: `uuid`,
    id: 59315,
    multiple: true,
    name: `scheduledService`,
    prototypeId: 59314,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.FreeForm,
    type: `uuid`,
  },
  serviceCost: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `serviceRecord`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `integer`,
    deepDataType: undefined,
    description: ``,
    displayName: `ServiceCost`,
    editable: false,
    format: { min: 1, max: 100, prefix: '$', suffix: 'k', examples: [] },
    formatType: `USD`,
    id: 59307,
    multiple: true,
    name: `serviceCost`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.NumericDiscrete,
    type: `integer`,
  },
  serviceDescription: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `serviceRecord`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `ServiceDescription`,
    editable: false,
    format: {
      examples: [
        'Flat Tire',
        'Broken Glass',
        'Oil Change',
        'Regular Service Interval',
        'Batttery',
        'Exterior Damage',
        'Engine Noise',
        'Suspension',
        'Brakes',
        'Electrical',
        'Interior',
        'Air Conditioning',
        'Navigation',
        'Error Code',
        'Emissions',
        'Inspection',
        'Accident Repair',
        'Transmission',
        'Performance',
        'Detailing',
      ],
    },
    formatType: `prime`,
    id: 59302,
    multiple: true,
    name: `serviceDescription`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  serviceRecord: {
    actions: [
      {
        areYouSure: true,
        component: 'ServiceRecordReportIssueModal',
        description: undefined,
        displayName: `Report Issue`,
        id: `4d882b7c-3773-443f-9327-16d168155e3c`,
        isCustomModal: true,
        name: `reportIssue`,
        type: ActionType.Custom,
      },
    ],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `dateScheduled`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceRecordType`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `downtimeHours`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `estimatedCost`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `datePerformed`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceCost`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceDescription`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `uuid`,
    deepDataType: undefined,
    description: ``,
    displayName: `ServiceRecord`,
    editable: false,
    format: { examples: [] },
    formatType: `uuid`,
    id: 59314,
    multiple: true,
    name: `serviceRecord`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.FreeForm,
    type: `uuid`,
  },
  serviceRecordType: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `serviceRecord`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `severity_level`,
    deepDataType: undefined,
    description: ``,
    displayName: `ServiceRecordType`,
    editable: false,
    format: {
      min: 1,
      max: 4,
      examples: ['Repair', 'Warranty', 'Recall', 'Preventative'],
      severityBad: 1,
      severityGood: 4,
      severityLevels: 4,
    },
    formatType: `severity_level`,
    id: 59313,
    multiple: true,
    name: `serviceRecordType`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.OrderedSeverityRanking,
    type: `severity_level`,
  },
  timestamp: {
    actions: [],
    attributes: [],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `timestamp`,
    deepDataType: undefined,
    description: undefined,
    displayName: `Timestamp`,
    editable: false,
    format: {},
    formatType: `timestamp`,
    id: 59271,
    multiple: true,
    name: `timestamp`,
    prototypeId: undefined,
    settings: {},
    statisticalType: StatisticalDataType.NumericTemporal,
    type: `timestamp`,
  },
  vehicle: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `vehicleModel`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `serviceRecord`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicleBrand`,
      },
      {
        hasMany: true,
        isNullable: false,
        name: `currentMileage`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `warrantyStatus`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicleWarranty`,
      },
      {
        hasMany: true,
        isNullable: false,
        name: `scheduledService`,
      },
      {
        hasMany: true,
        isNullable: false,
        name: `completedService`,
      },
    ],
    canAddValues: false,
    canEditValues: true,
    crossLinking: [{ title: 'Vehicle Detail', slug: 'vehicle-detail' }],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `Vehicle`,
    editable: true,
    format: {
      examples: [
        'Fleet Car 1',
        'Fleet Car 2',
        'Fleet Car 3',
        'Fleet Car 4',
        'Loaner Car',
        'Loaner Van',
        'Loaner Truck',
        'Service Van 1',
        'Service Van 2',
        'Delivery Van',
        'Delivery Truck',
        'Service Truck 1',
        'Service Truck 2',
        'Executive Car 1',
        'Executive Car 2',
        'Executive Car 3',
        'Sales Car 1',
        'Sales Car 2',
        'Sales Car 3',
      ],
    },
    formatType: `prime`,
    id: 59333,
    multiple: true,
    name: `vehicle`,
    prototypeId: undefined,
    settings: { isEditable: true, isFilledByEU: true },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  vehicleBrand: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `vehicle`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleBrand`,
    editable: false,
    format: {
      examples: [
        'Datsun',
        'Delorean',
        'Pontiac',
        'Plymouth',
        'Desoto',
        'Mercury',
        'Tucker',
        'Oldsmobile',
        'Studebaker',
        'Kaiser',
        'Saturn',
        'AMC',
        'Edsel',
        'Auburn',
        'Checker',
        'Dusenburg',
        'Fisker',
        'Hudson',
        'MGB',
        'Triumph',
      ],
    },
    formatType: `prime`,
    id: 59330,
    multiple: true,
    name: `vehicleBrand`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  vehicleDtc: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `vehicleDtcTimestamp`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: true,
        isNullable: false,
        name: `vehicleSoftwareVersion`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [{ title: 'DTC Detail', slug: 'dtc-detail' }],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleDtc`,
    editable: false,
    format: {
      examples: [
        'Air Cleaning Module Overcurrent - Circuit Current Above Threshold',
        'Backlite Defrost Relay Fault - Signal Shape / Waveform Failure',
        'Cabin Heater (PTCHF) 12V Logic Over Voltage - Circuit Voltage Above Threshold',
        'Door Handle Rear Left 12V Supply Under Voltage - Circuit Voltage Below Threshold',
        'Ext Door Handle Power RR Undercurrent - Circuit Current Below Threshold',
        'External 12VA Out of Range - Circuit Voltage Out of Range',
      ],
    },
    formatType: `prime`,
    id: 59726,
    multiple: true,
    name: `vehicleDtc`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  vehicleEcu: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicleSoftwareVersion`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleEcu`,
    editable: false,
    format: { examples: ['RIDM', 'BPCU', 'FIDM', 'RBCM', 'FBCM', 'DCAM', 'DCBM', 'OBCMD', 'OBCMP'] },
    formatType: `prime`,
    id: 48636,
    multiple: true,
    name: `vehicleEcu`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: true },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  vehicleModel: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `vehicle`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleModel`,
    editable: false,
    format: {
      examples: [
        'M1',
        'M2',
        'M3',
        'M5',
        'M4',
        'S1',
        'S2',
        'S3',
        'S4',
        'S5',
        'X1',
        'X2',
        'X3',
        'X4',
        'X5',
        'C1',
        'C2',
        'C3',
        'C4',
        'C5',
      ],
    },
    formatType: `prime`,
    id: 59328,
    multiple: true,
    name: `vehicleModel`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  vehicleSoftwareVersion: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `vehicleEcu`,
      },
      {
        hasMany: true,
        isNullable: true,
        name: `vehicleDtc`,
      },
      {
        hasMany: true,
        isNullable: false,
        name: `vehicle`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleSoftwareVersion`,
    editable: false,
    format: { examples: ['5.4.0', '5.4.1', '5.4.2', '5.5.0', '6.0.0'] },
    formatType: `prime`,
    id: 59725,
    multiple: true,
    name: `vehicleSoftwareVersion`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: true },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  vehicleWarranty: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `warrantyStatus`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `uuid`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleWarranty`,
    editable: false,
    format: { examples: [] },
    formatType: `uuid`,
    id: 59320,
    multiple: true,
    name: `vehicleWarranty`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.FreeForm,
    type: `uuid`,
  },
  vehicleStateOfCharge: {
    actions: [],
    attributes: [
      {
        hasMany: true,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
    ],
    canAddValues: false,
    canEditValues: true,
    crossLinking: [],
    dataType: `severity_score`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleStateOfCharge`,
    editable: true,
    format: { min: 0, max: 100, severityBad: 0, severityGood: 100, severityLevels: 5 },
    formatType: `severity_score`,
    id: 48640,
    multiple: true,
    name: `vehicleStateOfCharge`,
    prototypeId: undefined,
    settings: { isEditable: true, isFilledByEU: true },
    statisticalType: StatisticalDataType.NumericSeverityRanking,
    type: `severity_score`,
  },
  vehicleTimeOfLastCharge: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `vehicle`,
      },
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `string`,
    deepDataType: undefined,
    description: ``,
    displayName: `VehicleTimeOfLastCharge`,
    editable: false,
    format: { examples: ['02/23/2022, 4:28:57 PM', '05/23/2022, 4:28:57 PM'] },
    formatType: `prime`,
    id: 48633,
    multiple: true,
    name: `vehicleTimeOfLastCharge`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: true },
    statisticalType: StatisticalDataType.Categorical,
    type: `string`,
  },
  warrantyStatus: {
    actions: [],
    attributes: [
      {
        hasMany: false,
        isNullable: false,
        name: `timestamp`,
      },
      {
        hasMany: true,
        isNullable: false,
        name: `vehicleWarranty`,
      },
      {
        hasMany: true,
        isNullable: false,
        name: `vehicle`,
      },
    ],
    canAddValues: false,
    canEditValues: false,
    crossLinking: [],
    dataType: `severity_level`,
    deepDataType: undefined,
    description: ``,
    displayName: `WarrantyStatus`,
    editable: false,
    format: {
      min: 1,
      max: 2,
      examples: ['Expired', 'Active'],
      severityBad: 1,
      severityGood: 2,
      severityLevels: 2,
    },
    formatType: `severity_level`,
    id: 59319,
    multiple: true,
    name: `warrantyStatus`,
    prototypeId: undefined,
    settings: { isEditable: false, isFilledByEU: false },
    statisticalType: StatisticalDataType.OrderedSeverityRanking,
    type: `severity_level`,
  },
};
