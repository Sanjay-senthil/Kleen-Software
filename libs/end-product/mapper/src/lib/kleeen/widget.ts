import { CrossLinkingMatrix, DataAggregationArgsDataPoint, GetWidgetDataResult } from '@kleeen/types';

import { getThingByName } from '@kleeen/things';
import { isNilOrEmpty } from '@kleeen/common/utils';
import { path } from 'ramda';

export function mapToWidget(
  rows: any[],
  options: {
    categoryPath: string[];
    crossLinkPath?: string[];
    groupBy?: DataAggregationArgsDataPoint;
    resultsPath: string[];
    value?: DataAggregationArgsDataPoint;
  },
): GetWidgetDataResult {
  const categories = rows.map((row) => path(options.categoryPath, row));
  const xAxis = getThingByName(options?.groupBy?.name)?.format;
  const yAxis = getThingByName(options?.value?.name)?.format;

  function getCrossLinking(): CrossLinkingMatrix {
    const crossLinking = rows.map((row) => ({
      id: path(options.crossLinkPath, row),
      $metadata: {
        entityType: '',
      },
    }));

    return [crossLinking];
  }

  return {
    crossLinking: Array.isArray(options.crossLinkPath) ? getCrossLinking() : [],
    format: {
      xAxis: {
        ...xAxis,
        categories,
        key: options?.groupBy?.name,
        type: 'string',
      },
      yAxis: {
        ...yAxis,
        type: 'number',
        key: options?.value?.name,
      },
    },
    results: rows.map((row) => {
      const result = path(options.resultsPath, row);

      if (typeof result == 'number') return result;

      const isResultValidForNumberCoercion = !Array.isArray(result) || !isNilOrEmpty(result);

      return isResultValidForNumberCoercion ? Number(result) : 0;
    }),
  };
}
