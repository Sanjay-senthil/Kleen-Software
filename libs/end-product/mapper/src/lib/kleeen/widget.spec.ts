import { Maybe, Thing, Transformation } from '@kleeen/types';

import { mapToWidget } from './widget';
import { thingsMock } from './mocks';

jest.mock('@kleeen/things', () => {
  return {
    getThingByName(name: string): Maybe<Thing> {
      return Object.values(thingsMock).find((thing) => thing.name === name) as Thing;
    },
  };
});

describe('kleeen', () => {
  describe('widget', () => {
    it.todo(
      'should return "GetWidgetDataResult" response if "mapToWidget" is called from a Area Gradient, Step Line, Area Macro Micro, Area or a Simple List Ranked widget',
    );

    it.todo(
      'should return "GetWidgetDataResult" response if "mapToWidget" is called from a Area Trend widget',
    );

    it('should return "GetWidgetDataResult" response if "mapToWidget" is called from a Donut Variant, Donut, Pie or a Simple List Ranked widget', () => {
      const rows = [
        { 'Vehicle.warrantyStatus': 'Active', 'Vehicle.countTotal': 15 },
        { 'Vehicle.warrantyStatus': 'Expired', 'Vehicle.countTotal': 9 },
      ];
      const options = {
        categoryPath: ['Vehicle.warrantyStatus'],
        groupBy: { name: 'warrantyStatus', transformation: Transformation.SelfMulti },
        resultsPath: ['Vehicle.countTotal'],
        value: { name: 'vehicle', transformation: Transformation.CountTotal },
      };

      expect(mapToWidget(rows, options)).toMatchSnapshot();
    });

    it('should return "GetWidgetDataResult" response if "mapToWidget" is called from a List widget', () => {
      const rows = [
        { 'Service.category': 'Repair', 'Service.count': 108 },
        { 'Service.category': 'Warranty', 'Service.count': 10 },
        { 'Service.category': 'Preventative', 'Service.count': 9 },
        { 'Service.category': 'Recall', 'Service.count': 2 },
      ];
      const options = {
        categoryPath: ['Service.category'],
        groupBy: { name: 'serviceRecordType', transformation: Transformation.SelfMulti },
        resultsPath: ['Service.count'],
        value: { name: 'serviceRecord', transformation: Transformation.CountTotal },
      };

      expect(mapToWidget(rows, options)).toMatchSnapshot();
    });

    it('should return "GetWidgetDataResult" response if "mapToWidget" is called from a Polar Area or a Simple List Ranked widget', () => {
      const rows = [
        {
          'Dtc.id': 'B1021-96',
          'Dtc.name': 'Air Cleaning Module Overcurrent - Circuit Current Above Threshold',
          'Vehicle.count': '38',
        },
        {
          'Dtc.id': 'P1102-21',
          'Dtc.name': 'Backlite Defrost Relay Fault - Signal Shape / Waveform Failure',
          'Vehicle.count': '58',
        },
        {
          'Dtc.id': 'B1029-16',
          'Dtc.name': 'Cabin Heater (PTCHF) 12V Logic Over Voltage - Circuit Voltage Above Threshold',
          'Vehicle.count': '63',
        },
      ];
      const options = {
        categoryPath: ['Dtc.name'],
        crossLinkPath: ['Dtc.id'],
        groupBy: { name: 'vehicleDtc', transformation: Transformation.SelfMulti },
        resultsPath: ['Vehicle.count'],
        value: { name: 'vehicle', transformation: Transformation.CountTotal },
      };

      expect(mapToWidget(rows, options)).toMatchSnapshot();
    });

    it.todo(
      'should return "GetWidgetDataResult" response if "mapToWidget" is called from a Scatter or a Simple List Ranked widget',
    );

    it('should return "GetWidgetDataResult" response if "mapToWidget" is called from a Single-Bar Horizontal/Vertical or a Simple List Ranked widget', () => {
      const rows = [
        { 'Service.component': 'Battery', 'Service.totalCost': 193 },
        { 'Service.component': 'Brake', 'Service.totalCost': 179 },
        { 'Service.component': 'Coolant', 'Service.totalCost': 121 },
        { 'Service.component': 'Motor', 'Service.totalCost': 76 },
        { 'Service.component': 'Charging System', 'Service.totalCost': 51 },
        { 'Service.component': 'Navigation', 'Service.totalCost': 47 },
      ];
      const options = {
        categoryPath: ['Service.component'],
        groupBy: { name: 'serviceDescription', transformation: Transformation.SelfMulti },
        resultsPath: ['Service.totalCost'],
        value: { name: 'serviceCost', transformation: Transformation.CountTotal },
      };

      expect(mapToWidget(rows, options)).toMatchSnapshot();
    });
  });
});
