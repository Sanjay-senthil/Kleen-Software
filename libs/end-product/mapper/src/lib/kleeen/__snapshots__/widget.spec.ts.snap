// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kleeen widget should return "GetWidgetDataResult" response if "mapToWidget" is called from a Donut Variant, Donut, Pie or a Simple List Ranked widget 1`] = `
Object {
  "crossLinking": Array [],
  "format": Object {
    "xAxis": Object {
      "categories": Array [
        "Active",
        "Expired",
      ],
      "examples": Array [
        "Expired",
        "Active",
      ],
      "key": "warrantyStatus",
      "max": 2,
      "min": 1,
      "severityBad": 1,
      "severityGood": 2,
      "severityLevels": 2,
      "type": "string",
    },
    "yAxis": Object {
      "examples": Array [
        "Fleet Car 1",
        "Fleet Car 2",
        "Fleet Car 3",
        "Fleet Car 4",
        "Loaner Car",
        "Loaner Van",
        "Loaner Truck",
        "Service Van 1",
        "Service Van 2",
        "Delivery Van",
        "Delivery Truck",
        "Service Truck 1",
        "Service Truck 2",
        "Executive Car 1",
        "Executive Car 2",
        "Executive Car 3",
        "Sales Car 1",
        "Sales Car 2",
        "Sales Car 3",
      ],
      "key": "vehicle",
      "type": "number",
    },
  },
  "results": Array [
    15,
    9,
  ],
}
`;

exports[`kleeen widget should return "GetWidgetDataResult" response if "mapToWidget" is called from a List widget 1`] = `
Object {
  "crossLinking": Array [],
  "format": Object {
    "xAxis": Object {
      "categories": Array [
        "Repair",
        "Warranty",
        "Preventative",
        "Recall",
      ],
      "examples": Array [
        "Repair",
        "Warranty",
        "Recall",
        "Preventative",
      ],
      "key": "serviceRecordType",
      "max": 4,
      "min": 1,
      "severityBad": 1,
      "severityGood": 4,
      "severityLevels": 4,
      "type": "string",
    },
    "yAxis": Object {
      "examples": Array [],
      "key": "serviceRecord",
      "type": "number",
    },
  },
  "results": Array [
    108,
    10,
    9,
    2,
  ],
}
`;

exports[`kleeen widget should return "GetWidgetDataResult" response if "mapToWidget" is called from a Polar Area or a Simple List Ranked widget 1`] = `
Object {
  "crossLinking": Array [
    Array [
      Object {
        "$metadata": Object {
          "entityType": "",
        },
        "id": "B1021-96",
      },
      Object {
        "$metadata": Object {
          "entityType": "",
        },
        "id": "P1102-21",
      },
      Object {
        "$metadata": Object {
          "entityType": "",
        },
        "id": "B1029-16",
      },
    ],
  ],
  "format": Object {
    "xAxis": Object {
      "categories": Array [
        "Air Cleaning Module Overcurrent - Circuit Current Above Threshold",
        "Backlite Defrost Relay Fault - Signal Shape / Waveform Failure",
        "Cabin Heater (PTCHF) 12V Logic Over Voltage - Circuit Voltage Above Threshold",
      ],
      "examples": Array [
        "Air Cleaning Module Overcurrent - Circuit Current Above Threshold",
        "Backlite Defrost Relay Fault - Signal Shape / Waveform Failure",
        "Cabin Heater (PTCHF) 12V Logic Over Voltage - Circuit Voltage Above Threshold",
        "Door Handle Rear Left 12V Supply Under Voltage - Circuit Voltage Below Threshold",
        "Ext Door Handle Power RR Undercurrent - Circuit Current Below Threshold",
        "External 12VA Out of Range - Circuit Voltage Out of Range",
      ],
      "key": "vehicleDtc",
      "type": "string",
    },
    "yAxis": Object {
      "examples": Array [
        "Fleet Car 1",
        "Fleet Car 2",
        "Fleet Car 3",
        "Fleet Car 4",
        "Loaner Car",
        "Loaner Van",
        "Loaner Truck",
        "Service Van 1",
        "Service Van 2",
        "Delivery Van",
        "Delivery Truck",
        "Service Truck 1",
        "Service Truck 2",
        "Executive Car 1",
        "Executive Car 2",
        "Executive Car 3",
        "Sales Car 1",
        "Sales Car 2",
        "Sales Car 3",
      ],
      "key": "vehicle",
      "type": "number",
    },
  },
  "results": Array [
    38,
    58,
    63,
  ],
}
`;

exports[`kleeen widget should return "GetWidgetDataResult" response if "mapToWidget" is called from a Single-Bar Horizontal/Vertical or a Simple List Ranked widget 1`] = `
Object {
  "crossLinking": Array [],
  "format": Object {
    "xAxis": Object {
      "categories": Array [
        "Battery",
        "Brake",
        "Coolant",
        "Motor",
        "Charging System",
        "Navigation",
      ],
      "examples": Array [
        "Flat Tire",
        "Broken Glass",
        "Oil Change",
        "Regular Service Interval",
        "Batttery",
        "Exterior Damage",
        "Engine Noise",
        "Suspension",
        "Brakes",
        "Electrical",
        "Interior",
        "Air Conditioning",
        "Navigation",
        "Error Code",
        "Emissions",
        "Inspection",
        "Accident Repair",
        "Transmission",
        "Performance",
        "Detailing",
      ],
      "key": "serviceDescription",
      "type": "string",
    },
    "yAxis": Object {
      "examples": Array [],
      "key": "serviceCost",
      "max": 100,
      "min": 1,
      "prefix": "$",
      "suffix": "k",
      "type": "number",
    },
  },
  "results": Array [
    193,
    179,
    121,
    76,
    51,
    47,
  ],
}
`;
